### ISSUE TITLE
Feature request: support Yarn workspaces

### ISSUE THREAD

**Issue by eropple** [pulumi/pulumi#1868]
I am in a spot where I am building a bunch of different parts of a system all at once. I have a top-level _environment_ stack that pulls in _services_ defined by _components_--for example, the `infra` environment has a `gitlab` service with an `aws-rds-postgresql` component. Nothing crazy, I don't think, but each piece is being written to plug into the others.

Trying to do this is tricky with `npm` due to the need for changes in one Node module to be reflected in the next. The way I've been doing it so far is `npm link`, and that works okay-not-great, but now that I am two levels of linking deep, it is kind of a mess.

Typically I use `yarn` over `npm` mostly because `yarn` supports [workspaces](https://yarnpkg.com/en/docs/workspaces). When in a workspace, `yarn install` looks in the packages referenced by the top-level package (a private package) and links those rather than versions from a remote repository first. However, Pulumi fails when you have set up a workspace-based install (workspaces in the root package.json, installed via `yarn install`) with this error:

```
Diagnostics:
  global: global
    info: error: It looks like the Pulumi SDK has not been installed. Have you run npm install or yarn install?
 
error: failed to load language plugin nodejs: could not read plugin [/home/ed/.pulumi/bin/pulumi-language-nodejs] stdout: EOF
```

You also can't hack it together by linking `node_modules` into your "end" package (in my case, the environment package), without getting _this_ really fun error which is probably some other kind of nice bug:

```$ pulumi preview
error: failed to discover plugin requirements: re-statting file node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules: stat node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules/@carestack-cloud-environments/infra/node_modules: too many levels of symbolic links
```

If Pulumi could support workspaces, I think it'd probably make life a _lot_ easier for moderately complex problems.

**Comment by eropple** [comment]
**Workaround*: if you `yarn install` in the workspace, symbolically link its `node_modules` into your "end" project, and remove its own reference from `node_modules`, it does work, at least in my testing. So something like:

```
~/dev/pulumi $ yarn install
~/dev/pulumi $ cd environments/infra
~/dev/pulumi/environments/infra $ ln -s ../../node_modules
~/dev/pulumi/environments/infra $ rm node_modules/@carestack-cloud-environments/infra
```

Super hacky and not really reliable but usable for now.

**Comment by lukehoban** [comment]
Thanks - yes - we should make sure this works - it looks like the failures above are in places where we are trying to "guess" the package layout in ways that may not be compatible with Yarn workspaces.

Curious - where does `@pulumi/pulumi` end up resolving to in your setup?  The first failure above is because we expect it to be here (which seems to be too strict a check):
https://github.com/pulumi/pulumi/blob/fd852754fd358765ca26ffe5d9544c9ac06ceda0/sdk/nodejs/cmd/pulumi-language-nodejs/main.go#L59

It does appear that you may be able to set `PULUMI_LANGUAGE_NODEJS_RUN_PATH` to override this.

**Comment by eropple** [comment]
So it's in `$WORKSPACE_ROOT/node_modules/@pulumi/pulumi/cmd/run`, but if you can I would advise using `npx` or `yarn exec` to locate your deployables rather than attempting to guess it. There's a well-defined search hierarchy in Node and you can get it for free with `npx`/`yarn`.

If you can't use the default runners, you should walk up to the volume root looking for your file in progressively higher `node_modules` directories. See [the documentation](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) for details.

**Comment by ellismg** [comment]
Yarn workspaces aside, the logic we have is also totally bogus in cases where the `@pulumi/pulumi` module has been installed globally (e.g. `npm install -g @pulumi/pulumi`).

Ideally, I would like to move to a world where we just did something like `require("@pulumi/pulumi/cmd/run")` to load the module which would then run it.  I had hoped that we could do so via `node -e "..."`, but it looks like before 6.11.0 `node`, any arguments that started with a `-` when using `-e` were treated as node options.  In 6.11.0 they added support to allow `--`, and given that we were pushing 6.10.X for so long, I'm a little worried about taking that approach.

I'm still trying to understand the best way for us to fix this.  I had thought that perhaps we could have a `-exec` script side by side with `pulumi-language-nodejs` which did the require itself, but that also don't work because how node module resolution works (it starts the search from the folder of the -exec script, not the CWD).

It's possible we'll have to do something awful like the following (within `pulumi-language-nodejs`):

```
node -e 'try { console.log(require.resolve("@pulumi/pulumi/cmd/run")) } catch (e) { process.exit(1) }'
```

We could run that, capture the output and then use that as the path to run.

I am still a little worried, however, that there are more issues lurking, especially with how `run` itself uses require at run-time.  It also turns out that a bunch of the logic for detecting what plugins might be needed assume that there is a `node_modules` in a specific location, which will be incorrect (but fortunately, this is advisory information, so not getting a complete set of plugins should be okay).

Realistically, I don't think I'll have a solution I feel great about here over the next two days, so I'm moving this to M18.  Maybe some great idea will pop into my head.

**Comment by joeduffy** [comment]
Let's talk before moving out.

**Comment by eropple** [comment]
Is there a reason why you can't just make your `@pulumi/pulumi` runner into an executable that Node will then binstub? You can then just `npx` or `yarn exec` and be done with it.

Also, Node 6.x is out of maintenance LTS in like five months and is basically unpatched now. Node 8 seems to be the minimum anywhere I've been actively observing.

**Comment by lukehoban** [comment]
This is fixed with #1893, at least for the reported issue above.  

It's possible there are there are other places where we have problems here, and it would be nice to add some test coverage for non-standard NPM structures to identify those.  But closing this for now since all knowns issues are addressed.

**Comment by devdoshi** [comment]
I'm new to Pulumi but I am hitting the yarn v1 workspace issues with callbacks and dependencies not being included in the zip. I already have a working zip process for Serverless but I'd like it to be more reusable and efficient. I was thinking of making a node http server that is able to serve a yarn v1 workspace as a zip file. Does that seem like a useful workaround? Then I might make a custom Pulumi RemoteAsset based around that, which could maybe allow sending the parsed closure callback stuff into my server and have it put that in the right place in the zip.
